### PR TITLE
[Xtensa] Initial codegen support from IR

### PR DESCRIPTION
**Pull request by andreisfr**
This PR provides implementation of the basic codegen infra such as TargetFrameLowering, MCInstLower,
    AsmPrinter, RegisterInfo, InstructionInfo,  TargetLowering, SelectionDAGISel.

Migrated from https://reviews.llvm.org/D145658

### CODE REVIEW

**Comment by arsenm** [comment]
Unfortunately you should open a separate PR for each of these changes

**Review by s-barannikov — COMMENTED** [review]
Calling convention implementation needs to be in separate PR. It should support passing arguments on the stack, which requires implementing some methods in *FrameLowering, *InstrInfo and *RegisterInfo, as well as supporting load/store/call/return in ISel.

I would also suggest a separate PR for boilerplate classes, i.e. *RegisterInfo, *InstrInfo, *Subtarget, *FrameLowering, *TargetMachine, *AsmPrinter and *DAGToDAGISel. You can check that everything works by launching llc on an empty *.ll file (it should produce a template assembly (-filetype=asm) and object (-filetype=obj) files without crashing).

Supporting ALU operations should then touch only files related to instruction selection. (There may be other _minor_ changes in other files.)

Changes in MCTargetDesc/ look completely unrelated.

**Comment by andreisfr** [comment]
> Calling convention implementation needs to be in separate PR. It should support passing arguments on the stack, which requires implementing some methods in *FrameLowering, *InstrInfo and *RegisterInfo, as well as supporting load/store/call/return in ISel.
> 
> I would also suggest a separate PR for boilerplate classes, i.e. *RegisterInfo, *InstrInfo, *Subtarget, *FrameLowering, *TargetMachine, *AsmPrinter and *DAGToDAGISel. You can check that everything works by launching llc on an empty *.ll file (it should produce a template assembly (-filetype=asm) and object (-filetype=obj) files without crashing).
> 
> Supporting ALU operations should then touch only files related to instruction selection. (There may be other _minor_ changes in other files.)
> 
> Changes in MCTargetDesc/ look completely unrelated.

Thanks for the comment. Do I understand correctly that you are proposing to split this patch into three patches (the last one will be with ALU operations)?

**Comment by s-barannikov** [comment]
Yes, that's what I'm proposing.

**Comment by andreisfr** [comment]
> Yes, that's what I'm proposing.

I removed the Calling Convention functionality from this commit and will prepare a separate patch with it.

**Comment by andreisfr** [comment]
> Unfortunately you should open a separate PR for each of these changes



> Unfortunately you should open a separate PR for each of these changes

Thanks, I removed redundant commits.

File: llvm/lib/Target/Xtensa/XtensaSubtarget.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  HasDensity = false;
+
+  // Parse features string.
+  ParseSubtargetFeatures(CPUName, CPUName, FS);
+  return *this;
+}
+
+XtensaSubtarget::XtensaSubtarget(const Triple &TT, const std::string &CPU,
+                                 const std::string &FS, const TargetMachine &TM)
+    : XtensaGenSubtargetInfo(TT, CPU, /*TuneCPU*/ CPU, FS), TargetTriple(TT),
```
Comment (MaskRay) [inline llvm/lib/Target/Xtensa/XtensaSubtarget.cpp:44]: Canonical style is `/*TuneCPU=*/CPU`

**Comment by MaskRay** [comment]
`git log -- llvm/lib/Target/M68k/` M68k from 2021 sets a good example for new ports. You may follow it:)

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "XtensaGenInstrInfo.inc"
+
+namespace llvm {
+
+class XtensaTargetMachine;
+class XtensaSubtarget;
+class XtensaInstrInfo : public XtensaGenInstrInfo {
+  const XtensaRegisterInfo RI;
+  XtensaSubtarget &STI;
```
Comment (MaskRay) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.h:33]: Please ensure that each incremental change does not cause new warnings when compiled with a recent Clang.

You may use https://raw.githubusercontent.com/chromium/chromium/main/tools/clang/scripts/update.py  

```
warning: private field 'STI' is not used [-Wunused-private-field]
```

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+class MCContext;
+class MCInst;
+class MCOperand;
+class MCSymbol;
+class MachineInstr;
+class MachineOperand;
+class XtensaAsmPrinter;
+
+class LLVM_LIBRARY_VISIBILITY XtensaMCInstLower {
+  MCContext &Ctx;
```
Comment (MaskRay) [inline llvm/lib/Target/Xtensa/XtensaMCInstLower.h:29]: `warning: private field 'Ctx' is not used [-Wunused-private-field]`

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+SDValue XtensaTargetLowering::LowerOperation(SDValue Op,
+                                             SelectionDAG &DAG) const {
+  switch (Op.getOpcode()) {
+  default:
+    report_fatal_error("Unexpected node to lower");
+  }
+}
+
+const char *XtensaTargetLowering::getTargetNodeName(unsigned Opcode) const {
+  return NULL;
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:59]: nullptr

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+  // Reserve stack pointer.
+  Reserved.set(Xtensa::SP);
+  return Reserved;
+}
+
+bool XtensaRegisterInfo::eliminateFrameIndex(MachineBasicBlock::iterator II,
+                                             int SPAdj, unsigned FIOperandNum,
+                                             RegScavenger *RS) const {
+  report_fatal_error("Eliminate frame index not supported yet");
+  return false;
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:66]: dead return

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+Register XtensaRegisterInfo::getFrameRegister(const MachineFunction &MF) const {
+  const TargetFrameLowering *TFI = MF.getSubtarget().getFrameLowering();
+  return TFI->hasFP(MF) ? (Xtensa::A15) : Xtensa::SP;
```
Comment (arsenm) [inline llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp:71]: No () around register name.

Also should do getSubtarget<XTensaSubtarget>()

**Review by arsenm — COMMENTED** [review]
lgtm with some nits

**Comment by gerekon** [comment]
@MaskRay sorry for bothering you. I am from Espressif LLVM team.

> I think the patch series https://reviews.llvm.org/D145658 may be too fine-grained. It might be better to pack more changes in one PR, otherwise there could be a lot of `-Wunused-private-field` and not many testable units.

Looks like this conflicts with suggestions from @s-barannikov https://github.com/llvm/llvm-project/pull/78548#pullrequestreview-1837076213. As far as I understood he proposed to split it into more fine-grained commits.
Or I miss smth?

**Comment by arsenm** [comment]
> @MaskRay sorry for bothering you. I am from Espressif LLVM team.
> 
> > I think the patch series https://reviews.llvm.org/D145658 may be too fine-grained. It might be better to pack more changes in one PR, otherwise there could be a lot of `-Wunused-private-field` and not many testable units.
> 
> Looks like this conflicts with suggestions from @s-barannikov [#78548 (review)](https://github.com/llvm/llvm-project/pull/78548#pullrequestreview-1837076213). As far as I understood he proposed to split it into more fine-grained commits. Or I miss smth?

The split is better to review, but the pieces should build warning free and come with the relevant tests. I think this patch is fine with the warnings fixed

File: llvm/lib/Target/Xtensa/Xtensa.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
+//===- Xtensa.h - Top-level interface for Xtensa representation -*- C++ -*-===//
+//
+//                     The LLVM Compiler Infrastructure
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/Xtensa.h:3]: Current license header template doesn't include this string. I don't know if this is important or not.

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.h [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+namespace llvm {
+class MCContext;
+class MCInst;
+class MCOperand;
+class MCSymbol;
+class MachineInstr;
+class MachineOperand;
+class XtensaAsmPrinter;
+
+class LLVM_LIBRARY_VISIBILITY XtensaMCInstLower {
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaMCInstLower.h:28]: (optional) Is this class going to be large? If not, I'd propose to combine it with XtensaAsmPrinter. See https://github.com/llvm/llvm-project/commit/11e5a0d290bac87f3290547fa2b0aaffc98a798e

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.cpp [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/MachineInstr.h"
+#include "llvm/IR/Mangler.h"
+#include "llvm/MC/MCExpr.h"
+#include "llvm/MC/MCInst.h"
+#include "llvm/MC/MCStreamer.h"
+
+using namespace llvm;
+
+XtensaMCInstLower::XtensaMCInstLower(MCContext &ctx,
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaMCInstLower.cpp:27]: (nit) Variable names should start with a capital letter.

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.h [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+  const uint16_t *
+  getCalleeSavedRegs(const MachineFunction *MF = 0) const override;
+  const uint32_t *getCallPreservedMask(const MachineFunction &MF,
+                                       CallingConv::ID) const override;
+  BitVector getReservedRegs(const MachineFunction &MF) const override;
+  bool eliminateFrameIndex(MachineBasicBlock::iterator MI, int SPAdj,
+                           unsigned FIOperandNum,
+                           RegScavenger *RS = nullptr) const override;
+  Register getFrameRegister(const MachineFunction &MF) const override;
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaRegisterInfo.h:51]: (nit) I'd find it more readable with vertical spaces between method declarations.

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+} // end namespace llvm
+
+#endif /* LLVM_LIB_TARGET_XTENSA_REGISTERINFO_H */
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaRegisterInfo.h:56]: Use C++-style comments.

File: llvm/lib/Target/Xtensa/XtensaSubtarget.h [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+#define GET_SUBTARGETINFO_HEADER
+#include "XtensaGenSubtargetInfo.inc"
+
+namespace llvm {
+class StringRef;
+
+class XtensaSubtarget : public XtensaGenSubtargetInfo {
+private:
+  Triple TargetTriple;
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaSubtarget.h:35]:
```suggestion
  const Triple &TargetTriple;
```

File: llvm/lib/Target/Xtensa/XtensaSubtarget.h [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+
+  // Enabled Xtensa Density extension
+  bool HasDensity;
+
+  XtensaSubtarget &initializeSubtargetDependencies(StringRef CPU, StringRef FS);
+
+public:
+  XtensaSubtarget(const Triple &TT, const std::string &CPU,
+                  const std::string &FS, const TargetMachine &TM);
+
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaSubtarget.h:49]: const std::string & -> StringRef

TT parameter is redundant as it can be obtained from TM.

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h [hunk 14]
```diff
@@ -36,10 +37,15 @@ class XtensaTargetMachine : public LLVMTargetMachine {
                       std::optional<CodeModel::Model> CM, CodeGenOptLevel OL,
                       bool JIT);
 
+  const XtensaSubtarget *getSubtargetImpl() const { return &Subtarget; }
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaTargetMachine.h:40]: This method should not exist. Subtarget is per-function.

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h [hunk 15]
```diff
@@ -36,10 +37,15 @@ class XtensaTargetMachine : public LLVMTargetMachine {
...
+  const XtensaSubtarget *getSubtargetImpl(const Function &F) const override;
   TargetPassConfig *createPassConfig(PassManagerBase &PM) override;
   TargetLoweringObjectFile *getObjFileLowering() const override {
     return TLOF.get();
   }
+
+protected:
+  XtensaSubtarget Subtarget;
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaTargetMachine.h:48]: See the comment above, there should be no "default" subtarget.

File: llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+using namespace llvm;
+
+void XtensaAsmPrinter::emitInstruction(const MachineInstr *MI) {
+  XtensaMCInstLower Lower(MF->getContext(), *this);
+  MCInst LoweredMI;
+  Lower.lower(MI, LoweredMI);
+  EmitToStreamer(*OutStreamer, LoweredMI);
+}
+
+// Force static initialization.
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp:38]: The comment looks irrelevant.

**Review by s-barannikov — APPROVED** [review]
LGTM with comments in XtensaTargetMachine.h addressed.

File: llvm/lib/Target/Xtensa/XtensaSubtarget.cpp (hunk 1 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaSubtarget.cpp]: Thank you very much for comments. Fixed

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.h (hunk 2 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.h:33]: Fixed, I implemented "getSubtarget()" function.

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.h (hunk 3 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaMCInstLower.h]: Fxied.

File: llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+
+using namespace llvm;
+
+XtensaFrameLowering::XtensaFrameLowering()
+    : TargetFrameLowering(TargetFrameLowering::StackGrowsDown, Align(4), 0,
+                          Align(4)) {}
+
+// hasFP - Return true if the specified function should have a dedicated frame
+// pointer register.  This is true if the function has variable sized allocas or
+// if frame pointer elimination is disabled.
```
[the remark this answers is not in the document]
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaFrameLowering.cpp]: Thank you very much for comments. Fixed

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 4 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fxied

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp (hunk 5 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp (hunk 6 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaRegisterInfo.cpp]: Fixed

File: llvm/lib/Target/Xtensa/Xtensa.h (hunk 7 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/Xtensa.h]: Thank you very much for comments. I removed these strings from all new files.

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.h (hunk 8 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaMCInstLower.h]: Yes, this class not so large, so I combined it with XtensaAsmPrinter.

File: llvm/lib/Target/Xtensa/XtensaMCInstLower.cpp (hunk 9 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaMCInstLower.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.h (hunk 10 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaRegisterInfo.h:51]: Fixed

File: llvm/lib/Target/Xtensa/XtensaRegisterInfo.h (hunk 11 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaRegisterInfo.h]: Fixed

File: llvm/lib/Target/Xtensa/XtensaSubtarget.h (hunk 12 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaSubtarget.h]: Fixed

File: llvm/lib/Target/Xtensa/XtensaSubtarget.h (hunk 13 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaSubtarget.h]: Fixed

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h (hunk 14 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaTargetMachine.h]: Fixed. I implemented "getSubtargetImpl()" as it is done in M68k backend.

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h (hunk 15 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaTargetMachine.h]: Fxied

File: llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp (hunk 16 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaAsmPrinter.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+#define DEBUG_TYPE "xtensa-lower"
+
+XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &tm,
+                                           const XtensaSubtarget &STI)
+    : TargetLowering(tm), Subtarget(STI) {
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaISelLowering.cpp:33]:
```suggestion
XtensaTargetLowering::XtensaTargetLowering(const TargetMachine &TM,
                                           const XtensaSubtarget &STI)
    : TargetLowering(TM), Subtarget(STI) {
```

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+#include "llvm/CodeGen/MachineInstrBuilder.h"
+#include "llvm/CodeGen/MachineRegisterInfo.h"
+
+#define GET_INSTRINFO_CTOR_DTOR
+#include "XtensaGenInstrInfo.inc"
+
+using namespace llvm;
+
+XtensaInstrInfo::XtensaInstrInfo(XtensaSubtarget &sti)
+    : XtensaGenInstrInfo(), RI(sti), STI(sti) {}
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp:27]:
```suggestion
XtensaInstrInfo::XtensaInstrInfo(XtensaSubtarget &STI)
    : XtensaGenInstrInfo(), RI(STI), STI(STI) {}
```

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.h [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+
+public:
+  XtensaInstrInfo(XtensaSubtarget &STI);
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaInstrInfo.h:36]: The parameter should probably be `const`.

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h [hunk 21]
```diff
@@ -36,10 +37,14 @@ class XtensaTargetMachine : public LLVMTargetMachine {
                       std::optional<CodeModel::Model> CM, CodeGenOptLevel OL,
                       bool JIT);
 
+  const XtensaSubtarget *getSubtargetImpl(const Function &F) const override;
   TargetPassConfig *createPassConfig(PassManagerBase &PM) override;
   TargetLoweringObjectFile *getObjFileLowering() const override {
```
Comment (s-barannikov) [inline llvm/lib/Target/Xtensa/XtensaTargetMachine.h:42]: Could also use vertical spacing.

**Review by s-barannikov — APPROVED** [review]
LGTM, final style nits

File: llvm/lib/Target/Xtensa/XtensaSubtarget.h (hunk 13 above)
Comment (s-barannikov) [reply llvm/lib/Target/Xtensa/XtensaSubtarget.h]: Yes, but this is a very minor issue. Most backends pass it anyway.

File: llvm/lib/Target/Xtensa/XtensaISelLowering.cpp (hunk 18 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaISelLowering.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp (hunk 19 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaInstrInfo.cpp]: Fixed

File: llvm/lib/Target/Xtensa/XtensaTargetMachine.h (hunk 21 above)
Comment (andreisfr) [reply llvm/lib/Target/Xtensa/XtensaTargetMachine.h]: Fxied

**Review by MaskRay — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

**Comment by MaskRay** [comment]
I'll land this patch on your behalf tomorrow, unless someone objects it, assuming that you do not have write access yet!

**Comment by andreisfr** [comment]
> I'll land this patch on your behalf tomorrow, unless someone objects it, assuming that you do not have write access yet!

Thank you!

**Comment by gerekon** [comment]
[Tracking upstream PR in Espressif repo](https://github.com/espressif/llvm-project/issues/4)
